recon/src/eventid.rs

```diff
@@ -0,0 +1,177 @@
+#![warn(missing_docs, missing_debug_implementations, clippy::all)]
+
+use cbor::Encoder;
+use cid::Cid;
+use multihash::{Hasher, Sha2_256};
+use unsigned_varint::encode::u64 as varint;
+/// EventId generates EventIDs from event data.
+///
+/// varint(0xce) + // streamid, 1 byte
+/// varint(0x05) + // cip-124 EventID, 1 byte
+/// varint(networkId), // 1 byte (5 for local network)
+/// last8Bytes(sha256(separator_value)), // 16 bytes
+/// last8Bytes(sha256(stream_controller_DID)), // 16 bytes
+/// last4Bytes(init_event_CID) // 8 bytes
+/// cbor(eventHeight), // 1-3 bytes
+/// eventCID // 36 bytes
+///   0x01 cidv1, 1 byte
+///   0x71 dag-cbor, 1 byte
+///   0x12 sha2-256, 1byte
+///   0x20 varint(hash length), 1 byte
+///   hash bytes, 32 bytes
+
+/// Network values from https://cips.ceramic.network/tables/networkIds.csv
+/// Ceramic Pubsub Topic, Timestamp Authority
+#[derive(Debug)]
+pub enum Network {
+    /// /ceramic/mainnet Ethereum Mainnet (EIP155:1)
+    Mainnet,
+
+    /// /ceramic/testnet-clay Ethereum Gnosis Chain
+    TestnetClay,
+
+    /// /ceramic/dev-unstable Ethereum Gnosis Chain
+    DevUnstable,
+
+    /// /ceramic/local-$(randomNumber) Ethereum by Truffle Ganache
+    Local(u32),
+
+    /// None
+    InMemory,
+}
+
+impl Network {
+    /// network.to_u64() to get the network as a u64
+    pub fn to_u64(&self) -> u64 {
+        match self {
+            // https://github.com/ceramicnetwork/CIPs/blob/main/tables/networkIds.csv
+            Network::Mainnet => 0x00,
+            Network::TestnetClay => 0x01,
+            Network::DevUnstable => 0x02,
+            Network::InMemory => 0xff,
+            Network::Local(id) => 0x01_0000_0000_u64 + u64::from(*id),
+        }
+    }
+}
+
+/// event_id builds a Vec<u8> with the event id data.
+pub fn event_id(
+    network_id: Network,
+    separator: &str,
+    controller: &str,
+    init: &Cid,
+    event_height: u64,
+    event_cid: &Cid,
+) -> Vec<u8> {
```

Comment:
Should we create an `EventId` type? And make this method the `EventId::new` method?

Reply:
done